Development practices/folder_structure.md

```diff
@@ -0,0 +1,286 @@
+├── version_converter.rb
+└── x_server
+    ├── add_detour_distance.rb
+    ├── add_driving_distance.rb
+    ├── base.rb
+    ├── build_polyline.rb
+    ├── corridor_search.rb
+    ├── get_detour_distance.rb
+    ├── get_driving_distance.rb
+    ├── get_geo_keywords.rb
+    ├── get_location_attributes.rb
+    ├── get_route.rb
+    └── leave_reachable.rb
+```
+
+</details>
+
+At some point this folder became a dumpall folder without rhyme or reason.
+There is also an issue of boundries: there are none. Each class is used from anywhere in the application and that can
+cause headaches.
+
+# Desired structure
+
+Here is a folder structure we should strive for
+
+```
+app
+├── controllers  # Rails 
+├── frontend     # webpacker
+├── domains      # custom
+├── models       # Rails
+├── views        # Rails
+└── web          # custom
+```
+
+Lets first talk about Rails specific folders (`controllers`, `views` and `models`).
+
+`views` folder needs to exist because it is hardcoded in the Rails itself (https://github.com/rails/rails/blob/main/railties/lib/rails/engine.rb#L610)
+
+`controllers` is there because of the way rails finds controller names (it appends `"Controller"` to the resource name) (https://github.com/rails/rails/blob/main/actionpack/lib/action_dispatch/http/request.rb#L88)
+
+`models` is a folder where we put all our ActiveRecord models. It is technically not necessary but we will talk about it in [Further Improvements](#further-improvements) chapter.
+
+
+# `domains` vs `web` folder
+`web` is for classes used in controller actions: forms, policies, serializers, queries and similar.
+`domains` is for business logic.
+
+## `web` folder
+`web` folder structure should follow your routes. Having these routes:
+
+```ruby
+namespace: :dasboard do
+  resources :projects
+end
+namespace :api do
+  namespace :v1 do
+    resources :projects
+  end
+end
+```
+
+it would have this folder structure:
+
+```
+web
+├── dashboard
+│   ├── projects
+│   │   ├── presenter.rb
+│   │   ├── decorator.rb
+│   │   ├── form.rb
+│   │   └── policy.rb
+├── api
+│   ├── v1
+│   │   ├── v1
+│   │   │   ├── projects
+│   │   │   │   ├── policy.rb
+│   │   │   │   ├── query.rb
+│   │   │   │   └── serialzier.rb
+```
+
+## `domains` folder
+`domains` folder should be structured into domains. 
+
+The main idea is to create cohesive groups of domain objects that are involved in the same business use-case(s). The same way we try to build classes that are loosely coupled but highly cohesive we should build our domains as well.
+Here is an example of domains folder:
+
+```
+domains
+├── exports
+│   ├── formatters
+│   ├── csvs
+│   └── spreadsheets
+├── external
+│   └── zapier.rb
+├── projects
+│   ├── counter.rb
+│   ├── mailer.rb
+│   └── presister.rb
+├── utils
+│   ├── cache.rb
+│   └── slug.rb
+├── application_decorator.rb
+├── application_mailer.rb
+├── application_policy.rb
+├── application_presenter.rb
+├── application_query.rb
+├── application_serializer.rb
```

Comment:
My idea here was to have what Rails convention has: application_thing.
application_controller, application_record etc. To have that it would need to be in a root folder

But what @nikajukic is also a nice approach